### PR TITLE
DFU and multi SD support

### PR DESCRIPTION
**Pull request by mrodem**
This PR primarily adds device firmware upgrade (DFU) and support for multiple SoftDevices (v2 and v3). The `multi-sd-support` branch has been merged into this branch (`dfu`), and can be deleted after `dfu` has been merged into master.

Summary:
- API for reading DFU zip files and performing DFU
- Support for multiple SoftDevices (v2 and v3)
- Support for requesting higher ATT_MTU
- Memory leak fixes
- Unit testing of JavaScript code with Jest (`npm test` to run tests)

### CODE REVIEW

File: api/adapter.js [hunk 1]
```diff
@@ -323,7 +340,7 @@ class Adapter extends EventEmitter {
             options = {
                 gap_enable_params: {
                     periph_conn_count: 1,
-                    central_conn_count: 7,
+                    central_conn_count: this._bleDriver.NRF_SD_BLE_API_VERSION <= 2 ? 3 : 7,
```
Comment (bihanssen) [inline api/adapter.js]: central_conn_count should be set to 7 for both v2 and v3

File: api/adapter.js [hunk 2]
```diff
@@ -332,7 +349,10 @@ class Adapter extends EventEmitter {
                 },
                 common_enable_params: {
                     conn_bw_counts: null, // tell SD to use default
-                    vs_uuid_count: 10,
+                    vs_uuid_count: 5,
```
Comment (bihanssen) [inline api/adapter.js]: vs_uuid_count should still be 10

File: api/adapter.js [hunk 3]
```diff
@@ -1893,6 +1971,51 @@ class Adapter extends EventEmitter {
         });
     }
 
+    getCurrentAttMtu(deviceInstanceId) {
```
Comment (bihanssen) [inline api/adapter.js]: Note: add this function signature to the typings file

File: api/adapter.js [hunk 4]
```diff
@@ -2774,18 +2899,39 @@ class Adapter extends EventEmitter {
...
+        const MAX_ATTEMPTS = 10;
+        const RETRY_DELAY = 5;
+        const BLE_ERROR_NO_TX_PACKETS = 0x3004;
+        const tryWrite = () => {
+            this._adapter.gattcWrite(connectionHandle, writeParameters, err => {
+                if (err) {
+                    if (err.errno === BLE_ERROR_NO_TX_PACKETS && attempts++ <= MAX_ATTEMPTS) {
+                        setTimeout(() => tryWrite(), RETRY_DELAY);
+                    } else {
+                        callback(err);
```
Comment (bihanssen) [inline api/adapter.js]: Check if callback is defined

File: package.json [hunk 5]
```diff
@@ -18,15 +19,19 @@
         "bindings": "^1.2.1",
         "change-case": "2.3.0",
         "cmake-js": "1.1.0",
+        "crc": "^3.4.0",
         "nan": "2.3.3",
-        "underscore": "^1.8.3"
+        "underscore": "^1.8.3",
+        "jszip": "^3.1.2"
```
Comment (mrodem) [inline package.json]: Should order dependencies alphabetically.

File: src/driver.cpp [hunk 6]
```diff
@@ -1483,6 +1676,31 @@ ble_uuid128_t *BleUUID128::ToNative()
...
+
+ble_opt_t *BleOpt::ToNative()
+{
+    auto ble_opt = new ble_opt_t();
+    //memset(&ble_opt, 0, sizeof(ble_opt_t));
+
+    if (Utility::Has(jsobj, "common_opt")) {
+        //TODO: Implement common_opt
+        //auto common_opt_obj = ConversionUtility::getJsObjectOrNull(jsobj, "common_opt");
+        //ble_opt->common_opt = CommonOpt(common_opt_obj);
```
Comment (mrodem) [inline src/driver.cpp]: Some commented out code here. Is it something that we have forgotten to implement, or is it not needed yet?

File: test/dfu.js [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+
+const api = require('../index').api;
+const driver = require('../index').driver;
+
+const setup = require('./setup');
+const adapterFactory = setup.adapterFactory;
+
+function runTests(adapter) {
+    getManifestOK();
+//    getManifestNonExistingZipPath();
```
Comment (mrodem) [inline test/dfu.js]: This line can be removed.

File: test/dfu.js [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+function getManifestNonExistingZipPath() {
+    const dfu = new api.Dfu();
+    dfu.getManifest("non-existing/path.zip", (err, manifest) => {
+        if (err) {
+            console.log(err);
+        } else {
+            console.log(manifest);
+        }
+    });
+}
```
Comment (mrodem) [inline test/dfu.js]: This function can be removed.

File: test/dfu.js [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+        targetAddress: 'FC:EC:28:81:8B:84',
+        targetAddressType: 'BLE_GAP_ADDR_TYPE_RANDOM_STATIC',
+    };
+    const dfu = new api.Dfu('bleTransport', transportParameters);
+
+//    const zipPath = "./dfu/dfu_test_softdevice_bootloader_s132.zip";
+//    const zipPath = "./dfu/dfu_test_softdevice_s132.zip";
+//    const zipPath = "./dfu/dfu_test_bootloader_s132.zip";
+//    const zipPath = "./dfu/dfu_test_app_hrm_s132.zip";
+    const zipPath = "./dfu/dfu_test_all_s132.zip";
```
Comment (mrodem) [inline test/dfu.js]: Remove the commented out lines, and perhaps pass in the `zipPath` as a command line parameter.

File: test/dfu.js [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+                }
+            );
+        }
+    );
+}
+
+function listServices(adapter) {
+    const transportParameters = {
+        adapter: adapter,
+        targetAddress: 'FC:EC:28:81:8B:84',
```
Comment (mrodem) [inline test/dfu.js]: This line has to be edited by the person running the script. Perhaps pass the `targetAddress` as a command line parameter.

File: test/dfu.js [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+
+    adapter.on('logMessage', (severity, message) => { if(severity > 1) console.log(`#1 logMessage: ${message}`)});
+    adapter.on('status', (status) => { console.log(`#1 status: ${JSON.stringify(status)}`); });
+    adapter.on('error', error => { console.log('#1 error: ' + JSON.stringify(error, null, 1)); });
+    adapter.on('stateChanged', state => { console.log('#1 stateChanged: ' + JSON.stringify(state)); });
+    adapter.on('deviceDisconnected', device => { console.log('#1 deviceDisconnected: ' + JSON.stringify(device)); });
+    adapter.on('deviceDiscovered', device => { console.log(`Discovered device: ${JSON.stringify(device)}`); });
+    adapter.on('deviceConnected', device => {
+        console.log('#1 deviceConnected: ' + JSON.stringify(device));
+        deviceID = device._instanceId;
```
Comment (mrodem) [inline test/dfu.js]: The `deviceID` is not used. Can be removed.

File: test/dfu.js [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+            if (err) {
+                console.log('  performDFU failed: ', err);
+            }
+        });
+    });
+}
+
+adapterFactory.getAdapters((error, adapters) => {
+    assert(!error);
+    const adapter = adapters[Object.keys(adapters)[0]]
```
Comment (mrodem) [inline test/dfu.js]: Missing semicolon.

File: test/jest/dfu-test.js [hunk 13]
```diff
@@ -0,0 +1,257 @@
...
+
+const NRF_FAMILY = {
+    0: 'NRF51',
+    1: 'NRF52',
+};
+
+const CONNECTIVITY_HEX_FILES = {
+    0: './pc-ble-driver/hex/sd_api_v2/connectivity_1.0.1_115k2_with_s130_2.0.1.hex',
+    1: './pc-ble-driver/hex/sd_api_v3/connectivity_1.0.1_115k2_with_s132_3.0.hex',
+};
```
Comment (mrodem) [inline test/jest/dfu-test.js]: Update to connectivity 1.1.0.

File: test/dfu.js [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+        error => {
+            console.log("= inside adapter.connect callback")
+            if (error) {
+                console.log(error);
+            }
+            assert(!error);
+            if (callback) callback();
+        }
+    );
+}
```
Comment (mrodem) [inline test/dfu.js]: Connect is not needed. The DFU module connects if no connection exists.

File: api/adapter.js [hunk 15]
```diff
@@ -189,19 +206,19 @@ class Adapter extends EventEmitter {
...
     }
 
     _emitError(err, userMessage) {
-        let error = new Error(userMessage, err);
+        const error = new Error(userMessage, err);
         this.emit('error', error);
     }
 
-    _toHexString(value) {
+    static _toHexString(value) {
```
Comment (CoqRogue) [inline api/adapter.js]: This, together with _numberToxBitUuid, _arrayTo128BitUuid, etc should probably be moved to a separate utility library.

File: test/dfu.js (hunk 7 above)
Comment (mrodem) [inline test/dfu.js]: Done.

File: test/dfu.js (hunk 8 above)
Comment (mrodem) [inline test/dfu.js]: Done.

File: test/dfu.js (hunk 14 above)
Comment (mrodem) [inline test/dfu.js]: Done.

File: test/dfu.js (hunk 10 above)
Comment (mrodem) [inline test/dfu.js]: Now passing address as command line parameter.

File: test/dfu.js (hunk 9 above)
Comment (mrodem) [inline test/dfu.js]: Now passing zipPath as command line parameter.

File: test/dfu.js (hunk 11 above)
Comment (mrodem) [inline test/dfu.js]: Done.

File: test/dfu.js (hunk 12 above)
Comment (mrodem) [inline test/dfu.js]: Fixed.

File: test/jest/dfu-test.js (hunk 13 above)
Comment (mrodem) [inline test/jest/dfu-test.js]: Done.

File: package.json (hunk 5 above)
Comment (mrodem) [inline package.json]: Done.

File: api/adapter.js (hunk 1 above)
Comment (mrodem) [inline api/adapter.js]: Done.

File: api/adapter.js (hunk 2 above)
Comment (mrodem) [inline api/adapter.js]: Done.

File: api/adapter.js (hunk 3 above)
Comment (mrodem) [inline api/adapter.js]: Done.

File: api/adapter.js (hunk 4 above)
Comment (mrodem) [inline api/adapter.js]: Done.

File: api/adapter.js (hunk 15 above)
Comment (mrodem) [inline api/adapter.js]: Done. Moved these to `util/hexConv.js` and added some tests.

File: api/adapterFactory.js [hunk 16]
```diff
@@ -103,7 +107,39 @@ class AdapterFactory extends EventEmitter {
     _parseAndCreateAdapter(adapter) {
         // How about moving id generation and equality check within adapter class?
         const instanceId = this._getInstanceId(adapter);
-        const addOnAdapter = new this._bleDriver.Adapter();
+
+        let addOnAdapter; // = new this._bleDrivers.Adapter();
```
Comment (mrodem) [inline api/adapterFactory.js]: Done.

File: api/adapterFactory.js [hunk 17]
```diff
@@ -103,7 +107,39 @@ class AdapterFactory extends EventEmitter {
...
+        let selectedDriver;
+
+        // TODO: get adapters from one driver
+        // TODO: Figure out what device it is (nRF51 or nRF52). Perhaps serial number can be used?
+
```
Comment (bihanssen) [inline api/adapterFactory.js]: I'm don't think introducing a dependency to nrfjprog-js and jlinkarm dll for this small feature is worth it

File: api/adapter.js [hunk 18]
```diff
@@ -85,22 +85,20 @@ class Adapter extends EventEmitter {
...
         this._state = new AdapterState(instanceId, port, serialNumber);
-        this._security = Security.getInstance(this._bleDriver);
+        this._security = new Security(this._bleDriver);
         this._notSupportedMessage = notSupportedMessage;
 
-        this._maxReadPayloadSize = this._bleDriver.GATT_MTU_SIZE_DEFAULT - 1;
-        this._maxShortWritePayloadSize = this._bleDriver.GATT_MTU_SIZE_DEFAULT - 3;
-        this._maxLongWritePayloadSize = this._bleDriver.GATT_MTU_SIZE_DEFAULT - 5;
         this._keys = null;
+        this._attMtu = {};
```
Comment (bihanssen) [inline api/adapter.js]: done

File: api/adapter.js [hunk 19]
```diff
@@ -465,6 +485,9 @@ class Adapter extends EventEmitter {
                 case this._bleDriver.BLE_GATTC_EVT_TIMEOUT:
                     this._parseGattTimeoutEvent(event);
                     break;
+                case (this._bleDriver.NRF_SD_BLE_API_VERSION >= 3 ? this._bleDriver.BLE_GATTC_EVT_EXCHANGE_MTU_RSP : -1):
+                    this._parseGattcExchangeMtuResponseEvent(event);
+                    break;
```
Comment (bihanssen) [inline api/adapter.js]: Yes, I agree. I mistakenly assumed it would cause an error if the symbols were not defined, like in C.

Comment (bihanssen) [inline api/adapter.js]: done

File: api/dfu/dfuConstants.js [hunk 20]
```diff
@@ -0,0 +1,64 @@
+'use strict';
+
+// Control point operation codes for DFU BLE transport.
+// (Not to be confused with "NRF DFU Object codes".)
+const ControlPointOpcode = Object.freeze({
+    CREATE: 0x01, 0x01: 'Create',
```
Comment (bihanssen) [inline api/dfu/dfuConstants.js]: I vote for using a different way of doing reverse lookup of the objects. This way is kinda hackish and duplicates the info.

File: api/dfu/dfuModels.js [hunk 21]
```diff
@@ -0,0 +1,140 @@
...
+     * @param deviceState current state ({offset, crc32, maximumSize}) on device
+     */
+    constructor(data, deviceState) {
+        this._validateSize(data.length, deviceState.maximumSize);
+        this._deviceState = deviceState;
+        this._data = data;
+        this._hasResumablePartialObject = _canResumeWriting(this._data, deviceState.offset, deviceState.crc32);
+    }
+
+    get offset() {
```
Comment (bihanssen) [inline api/dfu/dfuModels.js]: Adapter.js uses the get keyword for getter functions. Consider using it here too.

File: api/dfu/dfuModels.js [hunk 22]
```diff
@@ -0,0 +1,140 @@
...
+    if (offset === 0 || offset > data.length || crc32 !== crc.crc32(data.slice(0, offset))) {
+        return false;
+    }
+    return true;
+}
+
+module.exports = {
+    InitPacketState,
+    FirmwareState,
+};
```
Comment (bihanssen) [inline api/dfu/dfuModels.js]: all files should have newline at end of file

File: api/dfu/dfuSpeedometer.js [hunk 23]
```diff
@@ -0,0 +1,61 @@
...
+        this._startTime = new Date();
+        this._lastUpdatedTime = this._startTime;
+        this._totalBytes = totalBytes;
+        this._completedBytes = completedBytes || 0;
+        this._initialBytes = completedBytes || 0;
+        this._percentRange = percentRange || { min: 0, max: 100 };
+        this._bytesPerSecond = 0;
+    }
+
+    getStartTime() {
```
Comment (bihanssen) [inline api/dfu/dfuSpeedometer.js]: consider get keyword for getter functions

File: api/util/intArrayConv.js [hunk 24]
```diff
@@ -0,0 +1,79 @@
...
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+'use strict';
+
+
+function intToArray(integer, sizeInBytes) {
+    // TODO: Check that integer is an integer.
```
Comment (bihanssen) [inline api/util/intArrayConv.js]: Add check and/or remove comment

File: index.js [hunk 25]
```diff
@@ -35,18 +35,18 @@
...
-var Adapter = require('./api/adapter');
-var AdapterFactory = require('./api/adapterFactory');
-var AdapterState = require('./api/adapterState');
-var Characteristic = require('./api/characteristic');
-var Descriptor = require('./api/descriptor');
-var Device = require('./api/device');
-var Security = require('./api/security');
-var Service = require('./api/service');
-var ServiceFactory = require('./api/serviceFactory');
+//var driver = require('bindings')('pc-ble-driver-js');
```
Comment (bihanssen) [inline index.js]: remove commented line

File: src/adapter.cpp [hunk 26]
```diff
@@ -133,21 +133,27 @@ void Adapter::initEventHandling(Nan::Callback *callback, uint32_t interval)
...
-        }
+    if (eventIntervalTimer == nullptr)
+    {
+        eventIntervalTimer = new uv_timer_t();
+    }
+
+    // Setup event interval functionality
+    eventIntervalTimer->data = static_cast<void *>(this);
+
+    if (uv_timer_init(uv_default_loop(), eventIntervalTimer) != 0) {
```
Comment (bihanssen) [inline src/adapter.cpp]: nitpick: brace to be placed on next line according to the project's c++ coding convention.

File: src/driver.cpp [hunk 27]
```diff
@@ -471,14 +496,14 @@ uint32_t Adapter::enableBLE(adapter_t *adapter)
...
     /* this application requires 1 connection as a peripheral */
     ble_enable_params.gap_enable_params.periph_conn_count = 1;
-    /* this application requires 3 connections as a central */
-    ble_enable_params.gap_enable_params.central_conn_count = 3;
     /* this application only needs to be able to pair in one central link at a time */
     ble_enable_params.gap_enable_params.central_sec_count = 1;
     /* we require the Service Changed characteristic */
     ble_enable_params.gatts_enable_params.service_changed = false;
-    /* the default Attribute Table size is appropriate for our application */
+    /* this application only needs to be able to pair in one central link at a time */
```
Comment (bihanssen) [inline src/driver.cpp]: this comment is misplaced

File: src/driver.cpp [hunk 28]
```diff
@@ -1204,6 +1237,160 @@ void Adapter::AfterReplyUserMemory(uv_work_t *req)
...
+    {
+        v8::Local<v8::String> message = ErrorMessage::getTypeErrorMessage(argumentcount, error);
+        Nan::ThrowTypeError(message);
+        return;
+    }
+
+    auto baton = new BleOptionBaton(callback);
+    baton->adapter = obj->adapter;
+    baton->opt_id = optionId;
+    baton->p_opt = new ble_opt_t();
```
Comment (bihanssen) [inline src/driver.cpp]: p_opt must be deleted in After-function to avoid memory leak

File: src/driver.cpp (hunk 6 above)
Comment (bihanssen) [inline src/driver.cpp]: It is not needed yet. 
Removed the lines, done.

File: api/dfu/dfuConstants.js (hunk 20 above)
Comment (mrodem) [inline api/dfu/dfuConstants.js]: Done in 6bb2a44.

File: src/driver.cpp (hunk 28 above)
Comment (bihanssen) [inline src/driver.cpp]: Done

File: api/util/intArrayConv.js (hunk 24 above)
Comment (bihanssen) [inline api/util/intArrayConv.js]: Done

File: index.js (hunk 25 above)
Comment (bihanssen) [inline index.js]: Done

File: api/dfu/dfuModels.js (hunk 22 above)
Comment (bihanssen) [inline api/dfu/dfuModels.js]: Done

**Pull request merged**